File: server/public/tools/elements/widgets/GtLabel.ts
import {Element, Listener} from "../../polymer/Annotations";
import {PolymerElement} from "../../polymer/PolymerElement";

/**
 * <gt-label>
 */
@Element({
	selector: "gt-label",
	template: "/assets/imports/widgets.html"
})
export class GtLabel extends PolymerElement {
	@Listener("click")
	private OnClick(ev: Event) {
		let control: any = this.node.querySelector("gt-input, gt-checkbox");
		if (control && control != ev.target) control.click();
	}

	@Listener("mouseenter")
	private OnEnter(ev: Event) {
		let control: any = this.node.querySelector("gt-input, gt-checkbox");
		if (control) control.setAttribute("hover", true);
	}

	@Listener("mouseleave")
	private OnLeave(ev: Event) {
		let control: any = this.node.querySelector("gt-input, gt-checkbox");
		if (control) control.removeAttribute("hover");
	}
}

File: client/src/gt/components/widget/form/GtLabel.scala
package gt.components.widget.form

import gt.components.GtHandler
import org.scalajs.dom._
import utils.implicits._
import utils.jsannotation.js
import xuen.Component

object GtLabel extends Component[GtLabel](
	selector = "gt-label",
	templateUrl = "/assets/imports/widgets.html"
) {
	val interactives = "gt-button, gt-checkbox, gt-input"
}

@js class GtLabel extends GtHandler {
	listen("click") { e: MouseEvent => forEachInteractives(i => i.click()) }
	listen("mouseenter") { e: MouseEvent => forEachInteractives(i => i.mouseenter()) }
	listen("mouseleave") { e: MouseEvent => forEachInteractives(i => i.mouseleave()) }

	def forEachInteractives(fn: Interactive => Unit): Unit = {
		for (item <- querySelectorAll(GtLabel.interactives)) fn(item.asInstanceOf[Interactive])
	}
}


